trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/testing/TestingS3RequestRewriter.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.aws.proxy.server.testing;
+
+import io.trino.aws.proxy.spi.credentials.Credentials;
+import io.trino.aws.proxy.spi.rest.ParsedS3Request;
+import io.trino.aws.proxy.spi.rest.S3RequestRewriter;
+
+import java.util.Optional;
+
+@FunctionalInterface
+public interface TestingS3RequestRewriter
```

Comment:
I think rather than having `TestingS3RequestRewriter` extending `S3RequestRewriter` we can have a `TestingS3RequestRewriteController` (we don't need a non-testing controller) that gets the bound `S3RequestRewriter` implementation and implements the `getTargetBucket` and `getTargetKey` utilities by delegating to that.

The main advantage is that we don't need a testing interface with 2 implementations. We can bind `TestingS3RequestRewriteController` as a concrete class. 
In test classes (like `TestProxiedRequestsWithRewrite`) we request the controller instead of the rewriter and we don't have to cast it.
In classes like `TestPresigningHeaders` we don't have to pass down the noop.

The NOOP implementation this was is handled by the `S3RequestRewriter` binding.

Reply:
Discussed this offline. `TestingS3RequestRewriter` is a helpful abstraction for tests since it is a simpler version of the "full" `S3RequestRewriter` - the former takes just a credential, bucket and key, while the latter expects a full `ParsedS3Request`.

If we got rid of the `TestingS3RequestRewriter` interface, we would need to construct dummy `ParsedS3Request`'s in each test to figure out where the request would be redirected to.

However, I agree the current casting of the request rewriter isn't ideal. We can create a request rewriter controller that looks a bit like what you suggested, and acts as a provider of `S3RequestRewriter`.

Installing `TestingS3RewriteController` would therefore result in the controller determining what rewriting logic gets used while providing a useful abstraction for tests